Optimization/tools

matteocacciola opened the pull request:
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context.  

# Checklist:

- [ ] I submitted my PR to branch `develop`
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have run or added tests to cover my contribution

matteocacciolaaubay commented on cat/looking_glass/bill_the_lizard.py:
```diff
@@ -264,10 +264,8 @@ async def replace_embedder(self, language_embedder_name: str, settings: Dict) ->
-
                 # inform the Cheshire Cats about the new embedder available in the system
-                await ccat.vector_memory_handler.initialize(self.embedder_name, self.embedder_size)
+                self.get_cheshire_cat(ccat_id)
```
the procedures are embedded in the constructor, and vector db is initialized therein

matteocacciolaaubay commented on cat/factory/vector_db.py:
```diff
@@ -70,15 +74,28 @@ def tenant_field_condition(self) -> Any:
+            if is_collection_existing and has_same_size:
+                continue
+
+            # dump collection on disk before deleting
+            if self.save_memory_snapshots:
+                await self.save_dump(collection_name)
+
+            if is_collection_existing:
+                await self.delete_collection(collection_name=collection_name)
+            await self.create_collection(embedder_name, embedder_size, collection_name)
```
`save_dump` is already an abstract method

matteocacciolaaubay commented on tests/test_agent.py:
```diff
@@ -50,13 +50,14 @@ async def mock_func(self, *args, **kwargs):
     monkeypatch.setattr("cat.mad_hatter.decorators.experimental.form.cat_form.CatForm.next", mock_func)
 
     # empty agent execution with form
+    tools = await stray.get_procedures()
```
`stray` has already an UserMessage

matteocacciola committed:
removing useless files
adjusting the README with the suppressed hooks
new version

matteocacciola committed:
minor changes and docstrings

matteocacciola committed:
starting with the optimization of the tool retrieval via procedural memory

matteocacciola committed:
optimizing the calling of procedures by using procedural memory

matteocacciola committed:
bugfix

matteocacciola committed:
fixing some tests

matteocacciola committed:
fixing test

matteocacciola committed:
reorganizing markdown files
final test fixing

matteocacciola committed:
moving elements to core plugins

matteocacciola committed:
adapting the hooks to run with lizard or cheshirecat or straycat
adapting the above classes to have an agent_key as a common interface
adjusting the March Hare plugin
adding some methods to the handler of the vector db, for hybrid searches
fixing tests
removing useless auth handler
